tests/test_app/test_install_plugins.py

```diff
@@ -1,256 +1,283 @@
-def install_temp_app(request: pytest.FixtureRequest):
-    """
-    A pytest fixture that creates a temporary app from a class, installs it,
-    and then uninstalls it after testing.
-
-    Parameters
-    ----------
-    request : pytest.FixtureRequest
-        The fixture request object. The class from which to create the app
-        should be passed as a parameter to the test function using this fixture.
-
-    Yields
-    ------
-    None
-
-    Raises
-    ------
-    TypeError
-        If the parameter passed to the test function is not a class.
-
-    Notes
-    -----
-    The app is installed in a temporary directory, which is deleted after testing.
-
-    Usage
-    -----
-
-    @pytest.mark.parametrize("install_temp_app", [MyAppClass], indirect=True)
-    def test_function(install_temp_app):
-        myapp = get_app('MyAppClass')
-        # test myapp
-    """
-    cls = request.param
-    if not isclass(cls):
-        raise TypeError(f"Expected a class, but got {type(cls).__name__}")
-
-    with temp_app(cls):
-        yield
-
-
-@pytest.mark.xfail(
-    reason="This test is expected to fail due to a bug with changing from using pkg_resources to importlib"
-)
-def test_install_temp_app():
-    @define_app
-    class MyAppClass:
-        def main(self, data: cogent3.app.typing.SerialisableType) -> str:
-            return json.dumps(data)
-
-        @classmethod
-        def _requires_imports(cls):
-            return ["from cogent3.app import typing as c3types", "import json"]
-
-    with temp_app(MyAppClass):
-        myapp = get_app("MyAppClass")
-        data = {"key": "value", "number": 42, "list": [1, 2, 3], "boolean": True}
-        assert (
-            myapp(data)
-            == '{"key": "value", "number": 42, "list": [1, 2, 3], "boolean": true}'
-        )
+#
+# The following code is used for dynamic app installation and uninstallation in tests
+#
+# The general structure of an app to be dynamically loaded is:
+#    @define_app
+#    class MyAppClass:
+#        def main(self, data: c3types.SerialisableType) -> str:
+#            return json.dumps(data)
+#        @classmethod
+#        def _requires_imports(cls):  # imports required by the app
+#            return ["from cogent3.app import typing as c3types", "import json"]
+#
+# Note: Due to the fact that it uses importlib, stevedore will see the first
+# installed app immediately, but subsequent apps may not be seen until the
+# importlib cache is eventually updated or the interpreter reloaded.
+# importlib.invalidate_caches() does not appear to work as expected.
+
+
+import importlib
+import json
+import os
+import site
+import sys
+import sysconfig
+import time
+
+from contextlib import contextmanager
+from importlib import metadata
+from inspect import getsourcelines
+from pathlib import Path
+from subprocess import check_call
+
+import pytest
+import stevedore
+
+from stevedore import ExtensionManager
+from stevedore._cache import _get_cache_dir
```

Comment:
OK, it's in a test so not so vulnerable for users if stevedore changes their API, which they may do without warning for private attributes... There must be a better way. Have you thought of posting a question on their repo on whether have a mechanism for supporting this type of evaluation?